docs/dev/guidelines/server.rst

```diff
@@ -269,7 +269,75 @@ you should use:
+    }
+
+Artemis distinguishes between five different roles: ADMIN, INSTRUCTOR, TA (teaching assistant), USER and ANONYMOUS.
+Each of the roles has the all the access rights of the roles following it, e.g. ANONOYMOUS has almost no rights, while ADMIN users can access every page.
+
+If a user passess the ``PreAuthorize`` check, the access to individual resources like courses and exercises still has to be checked. (A user can be a teaching assistant in one course, but only a student in another, for example.)
+However, do not fetch the user from the database yourself (unless you need to re-use the user object), but only hand a role to the ``AuthorizationCheckService``:
+
+.. code-block:: java
+
+        // If we pass 'null' instead of a user here, the service will fetch the user object
+        // and check if the user has at least the given role and access to the resource
+        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, exercise, null);
+
+To reduce duplication, do not add explicit checks for authorization or existence of an entity but always use the ``AuthorizationCheckService``:
+
+.. code-block:: java
+
+    @GetMapping(Endpoints.GET_FOR_COURSE)
+    @PreAuthorize("hasRole('TA')")
+    public ResponseEntity<List<ProgrammingExercise>> getActiveProgrammingExercisesForCourse(@PathVariable Long courseId) {
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        authCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.TEACHING_ASSISTANT, course, null);
+
+        List<ProgrammingExercise> exercises = programmingExerciseService.findActiveExercisesByCourseId(courseId);
+        return ResponseEntity.ok().body(exercises);
+    }
+
+
+The course repository call takes care of throwing a ``404 Not Found`` exception if there exists no matching course. The ``AuthorizationCheckService`` throws a ``403 Forbidden`` exception if the user with the given role is unauthorized. Afterwards delegate to a service or repository method. The code becomes much shorter, cleaner and more maintainable.
+
+
+21. Assert using the most specific overload method
+==================================================
+
+When expecting results use ``AsserThat`` for server tests. That call **must** be followed by another assertion statement like ``isTrue()``. It is best practice to use more specific assertion statement rather than always expecting boolean values.
+
+For example, instead of
+
+.. code-block:: java
+
+    assertThat(submissionInDb.isPresent()).isTrue();
+    assertThat(submissionInDb.get().getFilePath().contains("ffile.png")).isTrue();
+
+use the methods from inside the ``assertThat`` directly:
+
+.. code-block:: java
+
+    assertThat(submissionInDb).isPresent();
+    assertThat(submissionInDb.get().getFilePath()).contains("ffile.png");
+
+This gives better error messages when an assertion fails and improves the code readability. However, be aware that not all methods can be used for assertions like this.
```

Comment:
Maybe we could add that in some cases using `assertThat(....).as("errormessage").isTrue();` would be also an idea to get more information from a failing asserrt.
[->](https://stackoverflow.com/questions/28994316/can-you-add-a-custom-message-to-assertj-assertthat)

Reply:
Done.